src/v/cloud_roles/refresh_credentials.cc

```diff
@@ -179,14 +179,15 @@ cloud_roles::refresh_credentials::impl::calculate_sleep_duration(
 
 void cloud_roles::refresh_credentials::impl::increment_retries() {
     _retries += 1;
-    if (_retries > max_retries) {
-        vlog(clrl_log.error, "max retries reached, aborting");
-        throw std::runtime_error("max retries reached, aborting");
-    }
+    vassert(
+      _retries <= _retry_params.max_retries,
+      "retry count {} == max_retries {}, aborting",
```

Comment:
Yes, there are multiple new points in this PR where there is a hard stop of the redpanda process. and this introduces additional risk, some of these places are:

1. retry failed to cloud api enough times that we cannot recover.
2. cloud api returned bad response structure which we cannot extract credentials from.
3. we were unable to fetch the role for AWS EC2 deployment.
4. We were not able to fetch the required environment variables that AWS IRSA supplies us when running on k8s within AWS infra OR we do not have the token file supplied by k8s required to authenticate to IRSA (both the file and env variables should be present for us to be able to authenticate to STS)
5. communication errors to cloud APIs which are not recoverable such as 4xx responses.

We discussed this approach some time ago in SI war room, since any of these situations mean that effectively shadow indexing will not work, we do a hard stop because local storage will fill up and nothing will be uploaded to cloud storage. 

However if this needs further discussion because the risk introduced is high or this error handling is too aggressive, please let me know your thoughts. 

cc @Lazin